File: backend/app/api/v1/translate.py
"""
Translation API endpoints.

Provides POST /api/v1/translate endpoint for translating article content.
Requires JWT authentication via Authorization header.
Uses OpenAI API to translate content to Urdu.
"""

import logging
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
from openai import AsyncOpenAI

from app.db.models import User
from app.routes.auth import get_current_user, get_db
from app.schemas.translate import TranslateRequest, TranslateResponse
from app.core.config import settings

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/translate", tags=["translate"])

# Initialize OpenAI client
client = AsyncOpenAI(api_key=settings.OPENAI_API_KEY)


async def translate_text_openai(text: str, target_lang: str = "ur") -> str:
    """
    Translate text to Urdu using OpenAI API.

    Args:
        text: English text to translate
        target_lang: Target language code (currently only 'ur' for Urdu)

    Returns:
        Translated Urdu text

    Raises:
        Exception: If OpenAI API call fails
    """
    if target_lang != "ur":
        raise ValueError(f"Unsupported target language: {target_lang}")

    system_prompt = """You are a professional translator. Translate the provided English text to Urdu.
Important:
- Maintain the original meaning and tone
- PRESERVE ALL HTML TAGS EXACTLY as they appear (<h1>, <h2>, <h3>, <p>, <ul>, <li>, <strong>, <em>, <code>, <pre>, etc.)
- Only translate the text content INSIDE the tags, never modify or remove tags
- Preserve any formatting, code blocks, code placeholders [CODE_BLOCK_*], or special characters
- Translate only the text content, not code or technical terms where English is standard
- Keep the exact same HTML structure and layout
- Respond ONLY with the translated HTML, no explanations or additional content"""

    user_message = f"Translate the following English text to Urdu:\n\n{text}"

    try:
        response = await client.chat.completions.create(
            model=settings.OPENAI_MODEL,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_message},
            ],
            temperature=0.3,  # Lower temperature for more consistent translations
            max_tokens=min(len(text) * 2, 4000),  # Urdu text is typically longer
        )

        translated_text = response.choices[0].message.content.strip()
        logger.info(f"Successfully translated {len(text)} characters to Urdu")
        return translated_text

    except Exception as e:
        logger.error(f"OpenAI translation error: {str(e)}")
        raise


@router.post("/", response_model=TranslateResponse)
async def translate(
    request: TranslateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TranslateResponse:
    """
    Translate article content to target language.

    Requires valid JWT authentication token in Authorization header.

    Args:
        request: TranslateRequest with text to translate
        current_user: Authenticated user (guaranteed by get_current_user dependency)
        db: Database session (for potential future user tracking)

    Returns:
        TranslateResponse with translated text, language detection, and confidence

    Raises:
        HTTPException: 400 for invalid input, 401 for unauthenticated, 503 for service error
    """
    try:
        # Validate input
        if not request.text or len(request.text.strip()) == 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Text cannot be empty"
            )

        if len(request.text) > 50000:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Text too long (max 50000 chars)"
            )

        if request.target_lang not in ["ur"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Unsupported target language"
            )

        logger.info(f"Translation request from user {current_user.email}: {len(request.text)} chars to {request.target_lang}")

        # Call OpenAI API to translate
        translated_text = await translate_text_openai(
            text=request.text,
            target_lang=request.target_lang
        )

        return TranslateResponse(
            translated_text=translated_text,
            detected_lang=request.source_lang,
            confidence=0.95,  # OpenAI translations are highly reliable
            session_id=request.session_id
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Translation error: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Translation service error: {str(e)}"
        )

File: api/v1/translate.js
/**
 * POST /api/v1/translate - Translation endpoint
 *
 * Translates text to Urdu using OpenAI API.
 */

module.exports = async (req, res) => {
  // Enable CORS
  if (req.method === 'OPTIONS') {
    res.status(204).end();
    return;
  }

  if (req.method !== 'POST') {
    res.status(405).json({ detail: 'Method not allowed' });
    return;
  }

  try {
    const { text, target_lang, source_lang, session_id } = req.body || {};

    // Validate required fields
    if (!text || !text.trim()) {
      res.status(400).json({ detail: 'Text is required' });
      return;
    }

    if (!target_lang) {
      res.status(400).json({ detail: 'target_lang is required' });
      return;
    }

    // Only Urdu is supported
    if (target_lang !== 'ur') {
      res.status(400).json({ detail: 'Only Urdu (ur) translation is supported' });
      return;
    }

    // Check for OpenAI API key
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
      console.error('❌ OpenAI API key not configured');
      res.status(503).json({
        detail: 'Translation service not configured',
        code: 'service_error',
      });
      return;
    }

    // Call OpenAI API for translation
    const translatedText = await translateWithOpenAI(text, apiKey);

    res.status(200).json({
      translated_text: translatedText,
      detected_lang: source_lang || 'en',
      confidence: 0.95,
      session_id: session_id || null,
    });
  } catch (error) {
    console.error(`❌ Translate error: ${error.message}`);

    // Return appropriate error response
    if (error.code === 'auth_error') {
      res.status(401).json({ detail: 'Authentication failed' });
    } else if (error.code === 'validation_error') {
      res.status(400).json({ detail: error.message });
    } else {
      res.status(503).json({
        detail: 'Translation service error. Please try again later.',
        error: error.message,
      });
    }
  }
};

/**
 * Translate text to Urdu using OpenAI API
 */
async function translateWithOpenAI(text, apiKey) {
  const model = process.env.OPENAI_MODEL || 'gpt-4o-mini';

  const systemPrompt = `You are a professional translator. Translate the provided English text to Urdu.
Important:
- Maintain the original meaning and tone
- PRESERVE ALL HTML TAGS EXACTLY as they appear (<h1>, <h2>, <h3>, <p>, <ul>, <li>, <strong>, <em>, <code>, <pre>, etc.)
- Only translate the text content INSIDE the tags, never modify or remove tags
- Preserve any formatting, code blocks, code placeholders [CODE_BLOCK_*], or special characters
- Translate only the text content, not code or technical terms where English is standard
- Keep the exact same HTML structure and layout
- Respond ONLY with the translated HTML, no explanations or additional content`;

  const userMessage = `Translate the following English text to Urdu:\n\n${text}`;

  try {
    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model: model,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userMessage },
        ],
        temperature: 0.3,
        max_tokens: Math.min(text.length * 2, 4000),
      }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      console.error('OpenAI API error:', errorData);

      if (response.status === 401) {
        throw {
          code: 'auth_error',
          message: 'Invalid API key',
        };
      }

      throw {
        code: 'api_error',
        message: errorData.error?.message || 'OpenAI API error',
      };
    }

    const data = await response.json();
    const translatedText = data.choices[0]?.message?.content?.trim();

    if (!translatedText) {
      throw {
        code: 'validation_error',
        message: 'Empty response from translation service',
      };
    }

    return translatedText;
  } catch (error) {
    if (error.code) {
      throw error;
    }
    console.error('Translation fetch error:', error);
    throw {
      code: 'network_error',
      message: 'Failed to connect to translation service',
    };
  }
}
